neurots/generate/grower.py

```diff
@@ -367,6 +368,45 @@ def _grow_trunks(self):
                     )
                 )
 
+    def _3d_angles_grow_trunks(self):
+        """Grow trunk with 3d_angles method via :func:`.orientation.OrientationManager` class."""
+        trunk_orientations_manager = self._trunk_orientations_class(
+            soma=self.soma_grower.soma,
+            parameters=self.input_parameters,
+            distributions=self.input_distributions,
+            context=self.context,
+            rng=self._rng,
+        )
+        for neurite_type in self.input_parameters["grow_types"]:
+            orientations = trunk_orientations_manager.compute_tree_type_orientations(neurite_type)
+
+            for p in self.soma_grower.add_points_from_orientations(orientations):
+                self.active_neurites.append(
+                    TreeGrower(
+                        self.neuron,
+                        initial_direction=self.soma_grower.soma.orientation_from_point(p),
+                        initial_point=p,
+                        parameters=self.input_parameters[neurite_type],
+                        distributions=self.input_distributions[neurite_type],
+                        skip_validation=self.skip_validation,
+                        context=self.context,
+                        random_generator=self._rng,
+                    )
+                )
+
+    def _grow_trunks(self):
+        """Grow the trunks.
+
+        Two methods are available, depending on the data present in the `input_parameters`.
+        If no `3d_angles` entry is present, we grow trunks with :func:`_simple_grow_trunks` else
+        we fit the raw binned 3d angle data and apply :func:`_3d_angles_grow_trunks`.
+        """
+
+        if fit_3d_angles(self.input_parameters, self.input_distributions):
```

Comment:
You shouldn't mutate the attributes. They should be considered frozen. Make local copies if the initial data needs to be changed and pass it where needed. Debugging when the internal state of the instance changes based on how it is used is a nightmare.

I would dare go as far as converting the input_parameters and input_distribution into frozen dictionaries so that this is not allowed to occur :D

Reply:
so you suggest fit_3d_angles to return a copy of input_parameters? (it's the only one mutated)

Reply:
I don't quite like it, I prefered my clean way with a True/False return, but if you prefer, here it is